File: cpp-thin-client/src/ban_store.cpp
#include "ironwall/ban_store.hpp"
#include <chrono>

namespace ironwall {

static int64_t unix_now() {
    return std::chrono::duration_cast<std::chrono::seconds>(
        std::chrono::system_clock::now().time_since_epoch()).count();
}

BanStore::BanStore(std::string path) : path_(std::move(path)) {}

void BanStore::append_line(const std::string& line) {
    std::ofstream out(path_, std::ios::app);
    if (out) out << line << "\n";
}

void BanStore::save_ban(const std::string& player, const std::string& reason, double duration_sec) {
    int64_t until = unix_now() + static_cast<int64_t>(duration_sec);
    std::ostringstream ss;
    ss << "{\"op\":\"ban\",\"player\":\"" << player
       << "\",\"reason\":\"" << reason
       << "\",\"until_unix\":" << until << "}";
    append_line(ss.str());
}

void BanStore::save_unban(const std::string& player) {
    std::ostringstream ss;
    ss << "{\"op\":\"unban\",\"player\":\"" << player << "\"}";
    append_line(ss.str());
}

void BanStore::load_into(ModerationEngine& mod) {
    std::ifstream in(path_);
    if (!in) {
        std::cerr << "[ban_store] no file " << path_ << " (starting empty)\n";
        return;
    }
    std::string line;
    int loaded = 0;
    while (std::getline(in, line)) {
        if (line.find("\"op\":\"ban\"") != std::string::npos) {
            // crude extract player
            auto p0 = line.find("\"player\":\"");
            if (p0 == std::string::npos) continue;
            p0 += 10;
            auto p1 = line.find('"', p0);
            std::string player = line.substr(p0, p1 - p0);

            auto r0 = line.find("\"reason\":\"");
            std::string reason = "persisted";
            if (r0 != std::string::npos) {
                r0 += 10;
                auto r1 = line.find('"', r0);
                reason = line.substr(r0, r1 - r0);
            }

            auto u0 = line.find("\"until_unix\":");
            int64_t until = 0;
            if (u0 != std::string::npos) until = std::stoll(line.substr(u0 + 13));

            if (until > unix_now()) {
                mod.ban(player, reason + " (persisted)");
                ++loaded;
            }
        } else if (line.find("\"op\":\"unban\"") != std::string::npos) {
            auto p0 = line.find("\"player\":\"");
            if (p0 == std::string::npos) continue;
            p0 += 10;
            auto p1 = line.find('"', p0);
            mod.unban(line.substr(p0, p1 - p0));
        }
    }
    std::cerr << "[ban_store] loaded " << loaded << " active bans from " << path_ << "\n";
}

} // namespace ironwall

File: rust-thin-client/src/ban_store.rs
//! Persistent ban list (JSONL) — parity with C++ BanStore

use crate::moderation::ModerationEngine;
use chrono::Utc;
use std::fs::{File, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};
use tracing::info;

pub struct BanStore {
    path: PathBuf,
}

impl BanStore {
    pub fn new<P: AsRef<Path>>(path: P) -> Self {
        Self {
            path: path.as_ref().to_path_buf(),
        }
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    fn append(&self, line: &str) {
        if let Ok(mut f) = OpenOptions::new().create(true).append(true).open(&self.path) {
            let _ = writeln!(f, "{line}");
        }
    }

    pub fn save_ban(&self, player: &str, reason: &str, duration_secs: i64) {
        let until = Utc::now().timestamp() + duration_secs;
        self.append(&format!(
            r#"{{"op":"ban","player":"{player}","reason":"{reason}","until_unix":{until}}}"#
        ));
    }

    pub fn save_unban(&self, player: &str) {
        self.append(&format!(r#"{{"op":"unban","player":"{player}"}}"#));
    }

    pub fn load_into(&self, mod_eng: &mut ModerationEngine) {
        let f = match File::open(&self.path) {
            Ok(f) => f,
            Err(_) => {
                info!("ban_store: no file {} (empty)", self.path.display());
                return;
            }
        };
        let mut loaded = 0u32;
        for line in BufReader::new(f).lines().flatten() {
            if line.contains(r#""op":"ban""#) {
                let player = extract(&line, r#""player":""#).unwrap_or_default();
                let reason = extract(&line, r#""reason":""#).unwrap_or_else(|| "persisted".into());
                let until: i64 = line
                    .split(r#""until_unix":"#)
                    .nth(1)
                    .and_then(|s| s.split(|c: char| !c.is_ascii_digit() && c != '-').next())
                    .and_then(|s| s.parse().ok())
                    .unwrap_or(0);
                if until > Utc::now().timestamp() && !player.is_empty() {
                    mod_eng.ban(&player, &format!("{reason} (persisted)"));
                    loaded += 1;
                }
            } else if line.contains(r#""op":"unban""#) {
                if let Some(player) = extract(&line, r#""player":""#) {
                    mod_eng.unban(&player);
                }
            }
        }
        info!("ban_store: loaded {loaded} active bans from {}", self.path.display());
    }
}

fn extract(line: &str, key: &str) -> Option<String> {
    let i = line.find(key)? + key.len();
    let rest = &line[i..];
    let end = rest.find('"')?;
    Some(rest[..end].to_string())
}
